kafka/manager.go

```diff
@@ -146,83 +127,106 @@ func (m *Manager) DeleteTopics(ctx context.Context, topics ...apmqueue.Topic) er
 	return errors.Join(deleteErrors...)
 }
 
-func (m *Manager) gatherMetrics(ctx context.Context, o metric.Observer) error {
-	ctx, span := m.tracer.Start(ctx, "GatherMetrics")
-	defer span.End()
+// Healthy returns an error if the Kafka client fails to reach a discovered broker.
+func (m *Manager) Healthy(ctx context.Context) error {
+	if err := m.client.Ping(ctx); err != nil {
+		return fmt.Errorf("failed to ping kafka brokers: %w", err)
+	}
+	return nil
+}
 
-	// Fetch commits for consumer groups.
-	//
-	// TODO(axw) pass in a list of group names?
-	groups, err := m.adminClient.DescribeGroups(ctx)
-	if err != nil {
-		span.RecordError(err)
-		span.SetStatus(codes.Error, err.Error())
-		return fmt.Errorf("failed to describe groups: %w", err)
+// MonitorConsumerLag registers a callback with OpenTelemetry
+// to measure consumer group lag for the given topics.
+func (m *Manager) MonitorConsumerLag(topicConsumers []apmqueue.TopicConsumer) (metric.Registration, error) {
+	monitorTopicConsumers := make(map[apmqueue.TopicConsumer]struct{}, len(topicConsumers))
+	for _, tc := range topicConsumers {
+		monitorTopicConsumers[tc] = struct{}{}
 	}
-	consumerGroups := make([]string, 0, len(groups))
-	for _, group := range groups.Sorted() {
-		if group.ProtocolType != "consumer" {
-			m.cfg.Logger.Debug(
-				"ignoring non-consumer group",
-				zap.String("group", group.Group),
-				zap.String("protocol_type", group.ProtocolType),
-			)
-			continue
-		}
-		consumerGroups = append(consumerGroups, group.Group)
+
+	mp := m.cfg.meterProvider()
+	meter := mp.Meter("github.com/elastic/apm-queue/kafka")
+	consumerGroupLagMetric, err := meter.Int64ObservableGauge("consumer_group_lag")
+	if err != nil {
+		return nil, fmt.Errorf("kafka: failed to create consumer_group_lag metric: %w", err)
 	}
-	commits := m.adminClient.FetchManyOffsets(ctx, consumerGroups...)
-
-	// Fetch end offsets.
-	var endOffsets kadm.ListedOffsets
-	listPartitions := groups.AssignedPartitions()
-	listPartitions.Merge(commits.CommittedPartitions())
-	if topics := listPartitions.Topics(); len(topics) > 0 {
-		res, err := m.adminClient.ListEndOffsets(ctx, topics...)
+
+	gatherMetrics := func(ctx context.Context, o metric.Observer) error {
+		ctx, span := m.tracer.Start(ctx, "GatherMetrics")
+		defer span.End()
+
+		// Fetch commits for consumer groups.
```

Comment:
Can we pass in the consumer group names here? That will avoid listing all groups under the hood.

Reply:
Done!